gtsfm/frontend/correspondence_generator/mast3r_correspondence_generator.py

```diff
@@ -0,0 +1,341 @@
+                - idx2: Indices of correspondences in the second image's original feature grid.
+            """
+            image1_proc, crop1, scale1 = preprocess_image(image1, 0, device)
+            image2_proc, crop2, scale2 = preprocess_image(image2, 1, device)
+
+            with torch.no_grad():
+                model = model.to(device)
+
+                res = symmetric_inference(model, image1_proc, image2_proc, device=device)
+                X11, X21, X22, X12 = [r["pts3d"][0] for r in res]
+                C11, C21, C22, C12 = [r["conf"][0] for r in res]
+                descs = [r["desc"][0] for r in res]
+                qonfs = [r["desc_conf"][0] for r in res]
+
+                idx1, idx2, matches_im1, matches_im2, scores = extract_correspondences(
+                    descs, qonfs, device=device, subsample=subsample
+                )
+                conf_score = (C11.mean() * C12.mean() * C21.mean() * C22.mean()).sqrt().sqrt()
+                # TODO: use the matching score to filter correspondences.
+                matching_score = (float(conf_score), float(scores.sum()), len(scores))
+
+                matches_im1 = ((matches_im1 + np.array(crop1)) + 0.5) / scale1
+                matches_im2 = ((matches_im2 + np.array(crop2)) + 0.5) / scale2
+
+            return (matches_im1, matches_im2, idx1, idx2)
+
+        pairwise_correspondence_futures = {
+            (i1, i2): client.submit(
+                apply_mast3r,
+                m,
+                images[i1],
+                images[i2],
+            )
+            for i1, i2 in image_pairs
+        }
+        pairwise_correspondences: Dict[Tuple[int, int], Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]] = (
+            client.gather(pairwise_correspondence_futures)
+        )
+        logger.info(
+            "Mast3r computed correspondences for %d edges, aggregating them...", len(pairwise_correspondence_futures)
+        )
+
+        keypoints_for_image = {}
+        indices_for_image = {}
+        putative_corr_idxs_dict = {}
+
+        def update_keypoints(image_idx, keypoints, keypoints_idx):
+            """Helper function to update and deduplicate keypoints for a given image.
+
+            Args:
+                image_idx: The index of the image.
+                keypoints: The new keypoints to add.
+                keypoints_idx: The original indices of the new keypoints in the feature grid.
+            """
+            existing_idx = indices_for_image.get(image_idx, np.array([], dtype=np.int64))
+            indices_for_image[image_idx], unique_idx = np.unique(
+                np.concatenate([existing_idx, keypoints_idx]), return_index=True
+            )
+            existing_keypoints = keypoints_for_image.get(image_idx, np.array([], dtype=np.float32).reshape(0, 2))
+            keypoints_for_image[image_idx] = np.vstack([existing_keypoints, keypoints])[unique_idx]
```

Comment:
Would it be clearer to make this not a nested function?

Reply:
The other is designed to be a nested function (it changes those dicts inside it). I can force it to be non-nested if you prefer, but I dont think it hurts readability much as is.

the downside of being a non-nested function is that it becomes less obvious  that it is NOT functional and changes the dicts (so hurts readability IMO).

Reply:
That's makes sense